serdes-native/src/main/java/io/ballerina/stdlib/serdes/Serializer.java

```diff
@@ -0,0 +1,288 @@
+            String fieldName = entry.getKey().toString();
+            Object value = entry.getValue();
+            String unionType = fieldName + "_" + UNION_TYPE_IDENTIFIER;
+            Descriptor unionDescriptor = messageDescriptor.findNestedTypeByName(unionType);
+            if (unionDescriptor != null) {
+                DynamicMessage nestedMessage = generateDynamicMessageForBallerinaUnion(value, unionDescriptor,
+                                                                                       unionType);
+                newMessageFromSchema.setField(messageDescriptor.findFieldByName(fieldName), nestedMessage);
+            } else {
+                if (value == null) {
+                    continue;
+                }
+                if (value instanceof BMap) {
+                    BMap<BString, Object> objToBMap = (BMap<BString, Object>) value;
+                    String nestedTypeName = objToBMap.getType().getName();
+                    Descriptor subMessageDescriptor = schema.findNestedTypeByName(nestedTypeName);
+                    nestedTypeName = entry.getKey().toString();
+                    DynamicMessage nestedMessage = generateDynamicMessageForBallerinaRecord(objToBMap,
+                                                                                            subMessageDescriptor);
+                    newMessageFromSchema.setField(messageDescriptor.findFieldByName(nestedTypeName), nestedMessage);
+                } else if (value instanceof BArray) {
+                    FieldDescriptor field = messageDescriptor.findFieldByName(fieldName);
+                    generateDynamicMessageForBallerinaArray(newMessageFromSchema, schema, field, value, 1);
+                } else if (DataTypeMapper.getProtoTypeFromJavaType(value.getClass().getSimpleName()) != null) {
+                    FieldDescriptor field = messageDescriptor.findFieldByName(fieldName);
+                    generateDynamicMessageForBallerinaPrimitiveValue(newMessageFromSchema, field, value);
+                } else {
+                    String dataType = value.getClass().getSimpleName();
+                    throw createSerdesError(UNSUPPORTED_DATA_TYPE + dataType, SERDES_ERROR);
+                }
+            }
+        }
+        return newMessageFromSchema.build();
+    }
+
+    private static DynamicMessage generateDynamicMessageForBallerinaUnion(Object value, Descriptor schema,
+                                                                          String name) {
+        DynamicMessage.Builder newMessageFromSchema = DynamicMessage.newBuilder(schema);
+        Descriptor messageDescriptor = newMessageFromSchema.getDescriptorForType();
+
+        if (value == null) {
+            FieldDescriptor field = messageDescriptor.findFieldByName(NULL_FIELD_NAME);
+            newMessageFromSchema.setField(field, true);
+            return newMessageFromSchema.build();
+        }
+        String dataType = value.getClass().getSimpleName();
```

Comment:
Why do we need to access the class name here?

Reply:
To determine if the datatype is of a primitive type